test/regression/solvers/delta_eddington.cpp

```diff
@@ -0,0 +1,206 @@
+  ASSERT(accumulated_radiator_states.optical_depth_(0,0,1) > 2896168.4);
+  ASSERT(accumulated_radiator_states.optical_depth_(0,0,1) < 2896168.5);
+  ASSERT(accumulated_radiator_states.optical_depth_(83,41,1) > 4.83152e-4);
+  ASSERT(accumulated_radiator_states.optical_depth_(83,41,1) < 4.83153e-4);
+  ASSERT(accumulated_radiator_states.optical_depth_(155,119,1) > 6.64755e-10);
+  ASSERT(accumulated_radiator_states.optical_depth_(155,119,1) < 6.64756e-10);
+
+  // single scattering albedos are unitless
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_.Size1() == 156);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_.Size2() == 120);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_.Size3() == 2);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(0,0,0) > 4.78487e-6);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(0,0,0) < 4.78488e-6);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(83,41,0) > 0.5383450);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(83,41,0) < 0.5383451);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(155,119,0) > 0.997822);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(155,119,0) < 0.997823);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(0,0,1) > 4.78487e-6);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(0,0,1) < 4.78488e-6);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(83,41,1) > 0.5383450);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(83,41,1) < 0.5383451);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(155,119,1) > 0.997822);
+  ASSERT(accumulated_radiator_states.single_scattering_albedo_(155,119,1) < 0.997823);
+
+  // asymmetry parameters are unitless
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_.Size1() == 156);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_.Size2() == 120);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_.Size3() == 2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(0,0,0) > 2.12438e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(0,0,0) < 2.12439e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(83,41,0) > 2.13209e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(83,41,0) < 2.13210e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(155,119,0) == 0.0 );
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(0,0,1) > 2.12438e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(0,0,1) < 2.12439e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(83,41,1) > 2.13209e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(83,41,1) < 2.13210e-2);
+  ASSERT(accumulated_radiator_states.asymmetry_parameter_(155,119,1) == 0.0 );
+
+}
+
+SolverOutput RunDeltaEddingtonSolver(const SolverInput input)
```

Comment:
where do we define the test to call this function? I see the changes for the `CMakeLists.txt` file but do not see something like `test_xxx`.

Reply:
It's called by the fortran test driver, compiled here:
https://github.com/NCAR/tuv-x/blob/02388a93bc5a66c0a236271121beb274e1889401/test/regression/solvers/CMakeLists.txt#L9-L10

with an interface defined here:
https://github.com/NCAR/tuv-x/blob/02388a93bc5a66c0a236271121beb274e1889401/test/regression/solvers/delta_eddington.F90#L46-L51

and called here:
https://github.com/NCAR/tuv-x/blob/02388a93bc5a66c0a236271121beb274e1889401/test/regression/solvers/delta_eddington.F90#L211-L212

Reply:
@mattldawson So the driver is written in Fortran in this case? And I assume eventually we want to use google test to test the C++ solver?

Also is the Fortran version of delta Eddington solver being called as well for comparison? Or we compare the C++ output with some pre-existing Fortran output? I think you mentioned this to me before but sorry that I forgot the details.

Reply:
Yes, this test is just a regression test that runs the Fortran-based solver, copies all the solver input conditions to C++, calls the C++ solver, and then transfers the results back to Fortran to compare against the Fortran solver results.

But, we should definitely also create unit and integration tests in C++ with Google test for the C++ solver as it gets developed. If we can figure out some benchmark tests where we know the analytical solution, that would be ideal.

The only real reason for this test is that there were never any tests of the original Fortran code the solver is based on. It could be that there were bugs in the original code that will only show up through comparisons against the new, more tested implementation. Once we're convinced that the new solver is working and can explain any differences against the old solver, I think we can delete both this test and the Fortran solver source code.

Reply:
Thanks a lot for your detailed explanation. That is very helpful and clear to me now.